### DISCUSSION TITLE
atomFamilyWithStorage

### DISCUSSION THREAD

**Discussion by Cpt-Falcon** [pmndrs/jotai#2242] [Ideas]
Hey,

An idea that would be quite helpful for Jotai and give it a useful recoil feature would be an atomFamilyWithStorage. The atomFamilyWithStorage like with atomWithStorage would take in an identifier like 'dark mode', but it would also take in a unique identifier as well corresponding to each new atom in the family. Like a guid or an int id, whatever. Then, upon initialization in the atom family, the atom family would call a localStorageInitialization function. There it would look for the standard identifier like dark mode. It would scan through all local storage keys, looking if the string contained that dark mode identifier. If it does, we split the key string based on some kind of delimitator like a period. We would also take in a "onInitLoad" function from the user, the purpose being to parse and return the atom configured the way they want. The input to the atom would be the unique identifier as well as the content from local storage, and then would be expected to return their atom. 
 
I was going to try and implement this myself unless it already exists

**Comment by dai-shi** [comment]
I'm not exactly sure how it's different from `atomWithStorage` and `atomFamily` (or custom one) combination, but please feel free to give a try.

**Reply by Cpt-Falcon** [reply]
@dai-shi thanks for your time, I will submit a PR once I think I have met all the submission standards if you feel this suggestion has merit.  You're absolutely correct that logically speaking there isn't much of a difference, but I think the main benefit is that it could ultimately be a nice abstraction. atomFamilyWithStorage would automatically handle anything to do with the storage and retrieval for atoms contained in the map, and the entire atomFamily would be preloaded with the store. Right now, my understanding is that when you combine atomWithStorage and atomFamily, the atomFamily will only be populated once the identifier is accessed via initializeAtom in the atomFamily logic e.g. when its eventually accessed by some component in the application. If I was pulling from external storage, that could lead to slow downs if I wanted to frontload everything.  When I was a beginner, this nuance has led to some side effects in my code.

Furthermore, the storage and retrieval of an atom family, and all the atoms therein, could be fully contained within said atom family, possibly leading to cleaner, more reusable code instead of writing custom logic for each atom family. I don't know about others, but I almost always use atomWithStorage because its just a better experience when the app remembers the last state. Personally, such an abstraction would help to reduce code duplication in my projects.  

Admittedly, when I first started with Jotai I did not fully understand how to implement persistence with atom families or that atomWithStorage could be used in conjunction with atomFamily. I feel like an atomFamilyWithStorage would have made things very clear for beginners like me in the past. 
  
Anyways, thank you again for your time. I really can't emphasize enough how much my projects have benefitted from Jotai so i'd like to do whatever I can to help the community and this repo.

**Reply by dai-shi** [reply]
Please do note that I'm not confident if it should be in utils. My gut feeling is that if it's a simple util, it can be in a recipe, otherwise it's in a separate package. We will see. In the mean time, you can ideate your implementation here, or opening a draft PR is totally fine if you feel it's not extra work.

**Reply by mikeyzhong** [reply]
I've been having a world of trouble getting this to implement.

I have something

```
export const blueprintsAtomFamily = atomFamily(
  ({ id, blueprint }: BlueprintAndId) => {
    return atomWithStorage<Blueprint>(`blueprint-${id}`, blueprint);
  },
  (a, b) => a.id === b.id
);
```

It seems as though the atom is retrieved but it's not persisting in localStorage. Other examples of atomWithStorage without atomFamily wrapped around it work great on my end. Was wondering if someone here could shed some insight into what's going on or had a working example of atomFamily + atomWithStorage.

Thanks!

**Reply by dai-shi** [reply]
The code looks good though. You might want to provide a minimal repro to get a help from someone.

**Reply by mikeyzhong** [reply]
Thank you for the blazing fast response! I did some further digging and I think it may have something to do with this discussion https://github.com/pmndrs/jotai/discussions/2121, where state is not persisted in localStorage if the atom is never updated. Is there an easy way to create a dummy value for ```atomWithStorage<Blueprint>(`blueprint-${id}` , blueprint);``` so that the above line would cause an update instead of an initial state?

**Reply by dai-shi** [reply]
I'm not sure what you feel "easy", but `useEffect` should be one good option for this purpose.

**Reply by mikeyzhong** [reply]
Thanks!
